sdk/test/trace/tracer_test.cc

```diff
@@ -484,6 +495,137 @@ TEST(Tracer, SpanSetEvents)
+  EXPECT_EQ(typeid(span_foo_scope_with_version_ref), typeid(opentelemetry::sdk::trace::Span));
+
+  auto bar_scope = InstrumentationScope::Create("bar_library");
+  auto tracer_bar_scope =
+      initTracer(std::unique_ptr<SpanExporter>{new InMemorySpanExporter()}, new AlwaysOnSampler(),
+                 new RandomIdGenerator(), custom_configurator, std::move(bar_scope));
+  auto span_bar_scope      = tracer_bar_scope->StartSpan("span 1");
+  auto &span_bar_scope_ref = *span_bar_scope.get();
+  EXPECT_EQ(typeid(span_bar_scope_ref), typeid(opentelemetry::sdk::trace::Span));
+#else
+  GTEST_SKIP() << "cannot use 'typeid' with '-fno-rtti'";
+#endif
+}
+
+TEST(Tracer, StartSpanWithCustomConfigDifferingConditionOrder)
+{
+#ifdef OPENTELEMETRY_RTTI_ENABLED
+  auto check_if_version_present = [](const InstrumentationScope &scope_info) {
+    return !scope_info.GetVersion().empty();
+  };
+
+  // 2 configurators with same conditions, but different order
+  ScopeConfigurator<TracerConfig> custom_configurator_1 =
+      ScopeConfigurator<TracerConfig>::Builder(TracerConfig::Enabled())
+          .AddCondition(check_if_version_present, TracerConfig::Enabled())
+          .AddConditionNameEquals("foo_library", TracerConfig::Disabled())
+          .Build();
+  ScopeConfigurator<TracerConfig> custom_configurator_2 =
+      ScopeConfigurator<TracerConfig>::Builder(TracerConfig::Enabled())
+          .AddConditionNameEquals("foo_library", TracerConfig::Disabled())
+          .AddCondition(check_if_version_present, TracerConfig::Enabled())
+          .Build();
+
+  auto foo_scope_with_version_1 = InstrumentationScope::Create("foo_library", "1.0.0");
+  auto foo_scope_with_version_2 = InstrumentationScope::Create("foo_library", "1.0.0");
+
+  const auto tracer_foo_scope_with_version_1 = initTracer(
+      std::unique_ptr<SpanExporter>{new InMemorySpanExporter()}, new AlwaysOnSampler(),
+      new RandomIdGenerator(), custom_configurator_1, std::move(foo_scope_with_version_1));
+
+  const auto tracer_foo_scope_with_version_2 = initTracer(
+      std::unique_ptr<SpanExporter>{new InMemorySpanExporter()}, new AlwaysOnSampler(),
+      new RandomIdGenerator(), custom_configurator_2, std::move(foo_scope_with_version_2));
+
+  // Custom configurator 1 evaluates version first and enables the tracer
+  const auto span_foo_scope_with_version_1 = tracer_foo_scope_with_version_1->StartSpan("span 1");
+  auto &span_foo_scope_with_version_ref_1  = *span_foo_scope_with_version_1.get();
+  EXPECT_EQ(typeid(span_foo_scope_with_version_ref_1), typeid(opentelemetry::sdk::trace::Span));
+
+  // Custom configurator 2 evaluates the name first and therefore disables the tracer without
+  // evaluating other condition
+  const auto span_foo_scope_with_version_2 = tracer_foo_scope_with_version_2->StartSpan("span 1");
+  auto &span_foo_scope_with_version_ref_2  = *span_foo_scope_with_version_2.get();
+  EXPECT_EQ(typeid(span_foo_scope_with_version_ref_2), typeid(trace_api::NoopSpan));
```

Comment:
Possible alternative to using OPENTELEMETRY_RTTI_ENABLED, not tested, to investigate:

* Create a noop_span using NoopTracerProvider::GetTracer() and NoopTracer::StartSpan
* Compare the pointer to the span object with the noop span.

This relies on the fact that noop spans are always the same object, per:

```
    // Don't allocate a no-op span for every StartSpan call, but use a static
    // singleton for this case.
    static nostd::shared_ptr<trace::Span> noop_span(new trace::NoopSpan{this->shared_from_this()});

    return noop_span;
```

Alternate possibility:

* check for IsRecordable(), should be false for noop span and true for real spans.

Reply:
I like the idea of comparing to the noop_span directly. 

According to the spec, `isRecording()` might return true for a real span too - for instance if the span has ended. So I am preferring to use direct comparisons with the noop_span. 

One small caveat with this approach is that the test is now dependent on the implementation of the `NoopTracer`, but I feel it is OK for now.